src/ametsuchi/tx_store/iterator.cc

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2016 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <ametsuchi/tx_store/iterator.h>
+
+namespace ametsuchi {
+namespace tx_store {
+
+
+///////
+// Iterator
+
+Array::RandomAccessIterator::RandomAccessIterator(Array &array)
+    : array_(array), n_(0) {
+}
+
+Array::RandomAccessIterator::RandomAccessIterator(Array &array,
+                                                  std::size_t n)
+    :array_(array), n_(n)  {
+
+}
+
+Array::RandomAccessIterator::RandomAccessIterator(const RandomAccessIterator &it)
+    :array_(it.array_), n_(it.n_), value_(it.value_) {}
+
+ByteArray& Array::RandomAccessIterator::operator*() {
+  value_ = array_.get(n_);
+  return value_;
+}
+
+Array::RandomAccessIterator& Array::RandomAccessIterator::operator++() {
+  RandomAccessIterator iterator(*this);
+  n_ += 1;
+  return iterator;
```

Comment:
here you create *local* variable and return its *address*.
There will be undefined behavior.

Return copy, without `&`.

Read about "Return Value Optimization (RVO)"